cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -14,82 +19,151 @@ import qualified Data.Set as S
 import Distribution.Types.UnqualComponentName
-import Text.PrettyPrint (nest, render)
-
-data Message =
-    Enter           -- ^ increase indentation level
-  | Leave           -- ^ decrease indentation level
-  | TryP QPN POption
-  | TryF QFN Bool
-  | TryS QSN Bool
-  | Next (Goal QPN)
-  | Skip (Set CS.Conflict)
-  | Success
-  | Failure ConflictSet FailReason
-
--- | Transforms the structured message type to actual messages (strings).
+    ( unUnqualComponentName )
+
+import Text.PrettyPrint ( nest, render )
+
+renderSummarizedMessage :: SummarizedMessage -> String
+renderSummarizedMessage (SummarizedMsg i) = displayMessageAtLevel i
+renderSummarizedMessage (ErrorMsg s) = s
+
+displayMessageAtLevel :: EntryMsg -> String
+displayMessageAtLevel (AtLevel l msg) =
+  let s = show l
+  in  "[" ++ replicate (3 - length s) '_' ++ s ++ "] " ++ displayMessage msg
+
+displayMessage :: Entry -> String
+displayMessage (LogPackageGoal qpn gr) = "next goal: " ++ showQPN qpn ++ showGR gr
+displayMessage (LogRejectF qfn b c fr) = "rejecting: " ++ showQFNBool qfn b ++ showFR c fr
+displayMessage (LogRejectS qsn b c fr) = "rejecting: " ++ showQSNBool qsn b ++ showFR c fr
+displayMessage (LogSkipping cs) = "skipping: " ++ showConflicts cs
+displayMessage (LogTryingF qfn b) = "trying: " ++ showQFNBool qfn b
+displayMessage (LogTryingP qpn i mgr) = "trying: " ++ showQPNPOpt qpn i ++ maybe "" showGR mgr
+displayMessage (LogTryingS qsn b) = "trying: " ++ showQSNBool qsn b
+displayMessage (LogUnknownPackage qpn gr) = "unknown package" ++ showQPN qpn ++ showGR gr
+displayMessage LogSuccessMsg = "done"
+displayMessage (LogFailureMsg c fr) = "fail: " ++ showFR c fr
+displayMessage (LogSkipMany _ _ cs) = "skipping: " ++ showConflicts cs
+-- Instead of displaying `aeson-1.0.2.1, aeson-1.0.2.0, aeson-1.0.1.0, ...`,
+-- the following line aim to display `aeson: 1.0.2.1, 1.0.2.0, 1.0.1.0, ...`.
+--
+displayMessage (LogRejectMany qpn is c fr) = "rejecting: " ++ fmtPkgsGroupedByName (map (showQPNPOpt qpn) (reverse is)) ++ showFR c fr
+
+-- TODO: This function should take as input the Index? So even without calling the solver, We can say things as
+-- "There is no version in the Hackage index that match the given constraints".
+--
+-- Alternatively, by passing this to the solver, we could get a more semantic output like:
+-- `all versions of aeson available are in conflict with ...`. Isn't already what `tryToMinimizeConflictSet` is doing?
+fmtPkgsGroupedByName :: [String] -> String
+fmtPkgsGroupedByName pkgs = L.intercalate " " $ fmtPkgGroup (groupByName pkgs)
+  where
+    groupByName :: [String] -> M.Map String [String]
+    groupByName = foldr f M.empty
+      where
+        f versionString m = let (pkg, ver) = splitOnLastHyphen versionString
+                            in M.insertWith (++) pkg [ver] m
+        -- FIXME: This is not a very robust way to split the package name and version.
+        -- I should rather retrieve the package name and version from the QPN ...
```

Comment:
This is going to break on boot package names like `base-4.14.0.0/installed-4.14.0.0`.

Reply:
Yep, need to fix that.

Reply:
@Bodigrim Can you give an example of it breaking on boot package names? The `cabal-install:unit-tests` tests have instances of names like that and the tests pass fine.

Reply:
Please implement the rendered using structured types. Taking shortcuts is how bugs manifest.

Reply:
@mpickering Which types should I be using? I do not know this codebase well.

Reply:
Thanks @mpickering ! Changes made.